### DISCUSSION TITLE
Test with useButton gives me the famous "act" warning

### DISCUSSION THREAD

**Discussion by manelsanz** [adobe/react-spectrum#2743] [Q&A] [answered]
Hi, i'm using react-aria and react testing-library.
In my Button component, i'm using react-aria's `useButton` and I use `onPress` prop to pass the function to call once click. In the unit test, I just try to click it in order to trigger a simple jest.fn() function, and all I do after that is check if function have been called.

```
const handleOnClick = jest.fn();
render(<Button onPress={handleOnClick}>test me</Button>);

userEvent.click(screen.getByRole('button'));
expect(handleOnClick).toHaveBeenCalled();
```
That is showing the following warning:
```
console.error
    Warning: An update to Button inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):
```
If I wrap the click with act the warning is gone, but seems it's not a good practice to do that, because, in fact, testing library is doing an act internally in `userEvent` and  in other methods. What can I do? Thanks.

**Comment by snowystinger** [comment, marked as answer]
That test works for us https://github.com/adobe/react-spectrum/blob/c8381816b644e0a495106e3404637bc9e8aad084/packages/%40react-spectrum/button/test/Button.test.js#L34
Would you mind trying what we do in triggerPress? (we need to update that now that userEvent has had some releases that cover the problems we used to have with it)
https://github.com/adobe/react-spectrum/blob/main/packages/dev/test-utils/src/events.ts#L19
You could try some other versions of userEvent as well.

**Reply by manelsanz** [reply]
Thank you.  
I'm on last version of userEvent (13.5.0). I tried with `fireEvent.click` as you suggested, and it get rid of warning. But I would like to use `userEvent.click`, since it's behavior is more similar to the user and it's the recommeneded way to go. Even when, if I use `userEvent.click` with a normal `<button  onClick>` works as expected, without warning.

So, still need to find a solution..

**Reply by manelsanz** [reply]
Full warning message:
````javascript
 ✓ should execute the callback function onClick (139 ms)
  console.error
    Warning: An update to Button inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
        at Button (/home/USER/PROJECT/src/components/_global/atoms/Button/Button.tsx:21:26)

      at printWarning (node_modules/react-dom/cjs/react-dom.development.js:67:30)
      at error (node_modules/react-dom/cjs/react-dom.development.js:43:5)
      at warnIfNotCurrentlyActingUpdatesInDEV (node_modules/react-dom/cjs/react-dom.development.js:24064:9)
      at setPressed (node_modules/react-dom/cjs/react-dom.development.js:16135:9)
      at triggerPressEnd (node_modules/@react-aria/interactions/dist/packages/@react-aria/interactions/src/usePress.ts:179:7)
      at Document.onMouseUp (node_modules/@react-aria/interactions/dist/packages/@react-aria/interactions/src/usePress.ts:517:11)
      at Document.callTheUserObjectsOperation (node_modules/jsdom/lib/jsdom/living/generated/EventListener.js:26:30)
````

**Reply by snowystinger** [reply]
I just tried out the test and used userEvent.click instead of our triggerPress, we have userEvent version `@testing-library/user-event@12.1.3`
The test passes. Can you provide a codesandbox showing the test failing with the warning? or can you provide a git repo that reproduces it?

**Reply by manelsanz** [reply]
### Relevant library versions
````javascript
"react-aria": "3.12.0",
"react-dom": "17.0.2",
"react-scripts": "4.0.3",
"@testing-library/jest-dom": "5.16.1",
"@testing-library/react": "12.1.2",
"@testing-library/react-hooks": "7.0.2",
"@testing-library/user-event": "12.1.3",
````
Tried also with user-event 13.5.0 and 14.0.0.beta-8 versions  

### Minimum code to reproduce the warning (the test passes)

````javascript
import React, { useRef } from 'react';
import { useButton } from 'react-aria';
import userEvent from '@testing-library/user-event';

import { render, screen } from '@testing-library/react';

export const Button = (props: { onPress: () => void; children: React.ReactChild }) => {
  const ref = useRef<HTMLButtonElement>(null);
  const { buttonProps } = useButton(props, ref);
  const { children } = props;

  return (
    <button {...buttonProps} ref={ref}>
      {children}
    </button>
  );
};

describe('Atoms/Button', () => {
  const handleOnClick = jest.fn();

  it('should execute the callback function onClick', () => {
    render(<Button onPress={handleOnClick}>test me</Button>);

    const testButton = screen.getByRole('button');
    expect(testButton).toBeInTheDocument();
    expect(testButton).toHaveTextContent('test me');

    userEvent.click(testButton);

    expect(handleOnClick).toHaveBeenCalledTimes(1);
  });
});
````

### Full console output
````javascript
 PASS  src/components/_global/atoms/Button/tests/Button.test.tsx
  Atoms/Button
    ✓ should execute the callback function onClick (114 ms)

  console.error
    Warning: An update to Button inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
        at Button (/PROJECT_PATH/components/_global/atoms/Button/tests/Button.test.tsx:10:15)

      at printWarning (node_modules/react-dom/cjs/react-dom.development.js:67:30)
      at error (node_modules/react-dom/cjs/react-dom.development.js:43:5)
      at warnIfNotCurrentlyActingUpdatesInDEV (node_modules/react-dom/cjs/react-dom.development.js:24064:9)
      at setPressed (node_modules/react-dom/cjs/react-dom.development.js:16135:9)
      at triggerPressEnd (node_modules/@react-aria/interactions/dist/packages/@react-aria/interactions/src/usePress.ts:179:7)
      at Document.onMouseUp (node_modules/@react-aria/interactions/dist/packages/@react-aria/interactions/src/usePress.ts:517:11)
      at Document.callTheUserObjectsOperation (node_modules/jsdom/lib/jsdom/living/generated/EventListener.js:26:30)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        2.274 s
````

**Reply by snowystinger** [reply]
So I started a CRA app with typescript, installed the versions you mentioned, and copied the test you wrote into the App.test.tsx. It also runs with a green light and no warnings. Can you post a repo that demonstrates this?
Here's my app that I ran it in https://github.com/snowystinger/test-aria-button

**Reply by manelsanz** [reply]
You are right, sorry. Still I don't know what exactly is producing this warning, but I deleted package-lock.json (was in our repo) and node_modules folder and rerun `npm i`, and after deal with some `interopRequireDefault` error, could run tests without warning. Thank you and sorry for the disturbances. 

Have a good day!

**Reply by snowystinger** [reply]
no worries, happens some times, glad it's working now, that's all that matters :)

**Reply by pentzzsolt** [reply]
I am facing this same issue now. If you clone [my repository](https://github.com/pentzzsolt/very-good-metronome), run `npm install` and then `npm test`, the tests pass, but there is a warning:
```
  console.error
    Warning: An update to Button inside a test was not wrapped in act(...).
    
    When testing, code that causes React state updates should be wrapped into act(...):
    
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */
    
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
```
It's a simple app scaffolded using CRA Typescript template, and I added `@react-aria/button` as a dependency. The `Button` component looks like this:
```tsx
import { useButton } from '@react-aria/button';
import { ButtonProps } from './types/ButtonProps';
import { useRef } from 'react';

export const Button = ({ children, onPress }: ButtonProps): JSX.Element => {
  let ref = useRef(null);
  let { buttonProps } = useButton({
    onPress
  }, ref);

  return <button {...buttonProps} ref={ref}>{ children }</button>;
};
```
And the tests look like this:
```tsx
import { render, screen } from '@testing-library/react';
import { Button } from './Button';
import userEvent from '@testing-library/user-event';

const dummyButton = 'Dummy button';
const dummyFunction = jest.fn();

test('renders button with the correct label', () => {
  render(<Button>{ dummyButton }</Button>);
  const button = screen.getByRole('button', {
    name: dummyButton
  });
  expect(button).toBeVisible();
});

test('calls the function that was passed', () => {
  render(<Button onPress={dummyFunction}>{ dummyButton}</Button>);
  const button = screen.getByRole('button', {
    name: dummyButton
  });
  userEvent.click(button);
  expect(dummyFunction).toHaveBeenCalled();
});
```
Wrapping the `userEvent.click` call in `act` makes the warning disappear.

**Reply by LFDanLu** [reply]
From [user-event's installation page](https://testing-library.com/docs/user-event/install):

> If you use one of the [framework wrappers](https://testing-library.com/docs/dom-testing-library/install#wrappers), it is important that @testing-library/dom is resolved to the same installation required by the framework wrapper of your choice.
Usually this means that if you use one of the framework wrappers, you should not add @testing-library/dom to your project dependencies.

I installed your repo locally and noticed that multiple versions of that package were installed due to `@testing-library/react@13.4.0` and `@testing-library/user-event@13.5.0` having different deps for `@testing-library/dom`. I added the following override to your package.json and the tests ran without that warning:
```
  "overrides": {
    "@testing-library/dom": "8.20.1"
  },
```

**Reply by pentzzsolt** [reply]
Thanks for your reply! Interestingly, my repository is a new app scaffolded using CRA. It might be worth following up with them on this.
